walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,175 @@
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
+  /**
+   * Handles "find a meeting" feature by using the exsisting API to return the times
+   * when the meeting could happen that day, given the meeting information.
+   *
+   * @param events all events happening on a given day including event name, 
+   *                a TimeRange, and collection of attendees.
+   *
+   * @param request details of meeting being requested including requested meeting name, 
+   *                duration, and collection of attendees.
+   *
+   * @return an ArrayList containing all the TimeRanges the request meeting can be scheduled
+   *
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    } else if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<String> attendees = new ArrayList<String>(request.getAttendees());
+    List<TimeRange> busyTimeRanges = findBusyTimes(events, attendees);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
+      List<String> optional = new ArrayList<String>(request.getOptionalAttendees());
+      attendees.addAll(optional);
+
+      List<TimeRange> optionalBusyTimeRanges = findBusyTimes(events, attendees);
+      List<TimeRange> optionalMeetingTimes = findMeetingTimes(optionalBusyTimeRanges, duration);
+
+      if(!optionalMeetingTimes.isEmpty()) {
+        return optionalMeetingTimes;
+      } else {
+        return Arrays.asList();
+      }
+
+    }
+
+    return findMeetingTimes(busyTimeRanges, duration);
+  }
+
+  //finds meeting times that don't fall during busy times of the day
+  private List<TimeRange> findMeetingTimes(List<TimeRange> busyTimeRanges, long duration) {
+    if (busyTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(busyTimeRanges, TimeRange.ORDER_BY_START);
+
+    List<TimeRange> busyTimeMerged = mergeBusyTimes(busyTimeRanges);
+    List<TimeRange> sortByEndTime = sortByEnd(busyTimeMerged);
+
+    return findFreeTimes(busyTimeMerged, sortByEndTime, duration);
+  }
+
+  //merges all the overlapping busy times for all event attendees to create larger time interval, if possible
+  private List<TimeRange> mergeBusyTimes(List<TimeRange> busyTimeRanges) {
+    List<TimeRange> merged = new ArrayList<TimeRange>();
+
+    int numRange = busyTimeRanges.size();
+    if (numRange == 1) { //only one busy time range, no need to merge
+      return busyTimeRanges;
+    }
+ 
+    int currRangeIndex = 0;
+    int nextRangeIndex = 1;
+    int start = busyTimeRanges.get(0).start();
```

Comment:
I think time in time range should be long instead of int. Worth double checking that you are not doing a cast here.

Reply:
Oh that is surprising. I think start should be a time, right? busyTimeRanges.get(0) gives you a time range, and timeRange.start() gives you the start time.

Reply:
I think start is a number, cause it seems like, from the testing outputs, the times are all saved in number form, and TimeRange is the start and end number together [int, int], but I could also be wrong.

Reply:
Just checked the code's dependency, it is using a custom TimeRange class (that lives in the same directory). The custom TimeRamge class declares it as an int.

So you are good here. Sorry for the confusion :)